File: apple/LC98ValidateBinarySearchTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        return dfs(root, null, null);
    }

    public boolean dfs(TreeNode node, Integer minimum, Integer maximum) {
        if (node == null) return true;
        if (minimum != null && node.val <= minimum) return false;
        if (maximum != null && node.val >= maximum) return false;
        return dfs(node.left, minimum, node.val) && dfs(node.right, node.val, maximum);
    }
}

File: asml/LC98ValidateBinarySearchTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return dfs(root, NULL, NULL);
    }

    bool dfs(TreeNode* root, TreeNode* minimum, TreeNode* maximum) {
        if (root == NULL) return true;
        if (minimum != NULL && root->val <= minimum->val) return false;
        if (maximum != NULL && root->val >= maximum->val) return false;
        return dfs(root->left, minimum, root) && dfs(root->right, root, maximum);
    }
};
